README.md
# file_list_previewer

A flutter package for viewing a list of local files in the form of clickable thumbnails.

## Features:
* Covers different file types
* Presents the thumbnails for different types of files
* Preview of the file by clicking on the thumbnail 
* Dynamically Remove files from list and handle it back in the parent widget

### Getting Started

<img src="https://raw.githubusercontent.com/walkingtree/file_list_previewer/master/assets/images/2020-09-02.jpg" width="200" height="200">
<br />
<img src="https://raw.githubusercontent.com/walkingtree/file_list_previewer/master/assets/images/2020-09-02%20(2).jpg" width="200" height="200">
<br />
<img src="https://raw.githubusercontent.com/walkingtree/file_list_previewer/master/assets/images/2020-09-02%20(3).jpg" width="200" height="200">
<br />
<img src="https://raw.githubusercontent.com/walkingtree/file_list_previewer/master/assets/images/2020-09-02%20(4).jpg" width="200" height="200">
<br />
<img src="https://raw.githubusercontent.com/walkingtree/file_list_previewer/master/assets/images/2020-09-02%20(1).jpg" width="200" height="200">

#### Properties:
*//Either one of the attachmentList or filePaths should be passed*

```final List<File> attachmentList;```    *// the list of files to be displayed*<br />
```final List<String> filePaths;```       *// the list of local filepaths of files to be displayed*<br />
```final Function removeImage;```         *// An optional function parameter to handle file removal from the list of files/filepaths in parent widget*<br />
```final String videoFileImage;```        *//The thumbnail image for video thumbnail to be used instead of default* <br />
```final double width;```                 *//Pass width for the widget*<br />
```final double height;```                *//Pass height for the widget*<br />


##### Usage:

>   FileListPreviewer(
>              removeImage: _removeImage,
>              attachmentList: attachmentList,
>              filePaths: fileattachmentList,
>            ),

CHANGELOG.md
## [0.0.1] 

* Initial release.

## [0.0.2] 

* Removed unused dependency.
* Documentation added


test/file_list_previewer_test.dart
 import 'package:flutter/cupertino.dart';
import 'package:flutter_test/flutter_test.dart';

 import 'package:file_list_previewer/file_list_previewer.dart';

 void main() {
  testWidgets('file list preview', (WidgetTester tester)async{

   await tester.pumpWidget(FileListPreviewer());

   final singleChildScrollViewFinder  = find.byKey(ValueKey("singlechildkey"));
   expect(singleChildScrollViewFinder, findsOneWidget);

   final containerFinder  = find.byKey(ValueKey("containerkey"));
   expect(containerFinder, findsOneWidget);

   final listViewFinder  = find.byKey(ValueKey("listviewkey"));
   expect(listViewFinder, findsOneWidget);


  });

 }


example/lib/main.dart

import 'package:example/file_List_Preview.dart';
import 'package:file/file.dart';
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'ImageInput.dart';
import 'dart:io' as io;

void main() {
  runApp(MyApp(

  ));
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.green,
               visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {


  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<io.File> attachmentList = [];
  List<String> fileattachmentList = [];
  void _selectImage(io.File pickedImage, filename) {

    setState(() {

      attachmentList.add(pickedImage);
      fileattachmentList.add(pickedImage.path);
    });
  }

  void _removeImage(io.File pickedFile) {

    setState(() {
      attachmentList.remove(pickedFile);
      fileattachmentList.remove(pickedFile.path);
    });
  }

  @override
  Widget build(BuildContext context) {

    return Scaffold(
      appBar: AppBar(
          title:Text("File List Preview Example")
      ),
      body: Column(
       // crossAxisAlignment: CrossAxisAlignment.center,
//        mainAxisAlignment: MainAxisAlignment.center,
        children: <Widget>[
          Padding(
            padding: const EdgeInsets.all(20),
            child: Container(
              child: ImageInput(_selectImage),
            ),
          ),
          attachmentList.length >= 1
              ? Padding(
            padding: const EdgeInsets.all(5),
            child: FileListPreviewer(
              removeImage: _removeImage,
              attachmentList: attachmentList,
              filePaths: fileattachmentList,
            ),
          )
              : SizedBox(),
        ],
      ),
      // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


example/lib/ImageInput.dart
import 'dart:io';
import 'package:camera/camera.dart';
import 'package:example/TakePicturePage.dart';
import 'package:file_picker/file_picker.dart';
import 'dart:io' as io;
import 'package:flutter/material.dart';
import 'package:path/path.dart' as path;
import 'package:path_provider/path_provider.dart' as syspaths;

class ImageInput extends StatefulWidget {
  final Function onSelectImage;

  ImageInput(this.onSelectImage);

  @override
  _ImageInputState createState() => _ImageInputState();
}

class _ImageInputState extends State<ImageInput> {
  File _storedImage;
  String _filename;

  String fileName;

  var pickedFile;

  io.File savedImage;
  io.File attachFile;

  var showDropdown = false;
  void toggleFlag() {
    setState(() {
      showDropdown = !showDropdown;
    });
  }

  Future<void> _takePicture(_value) async {
    if (_value == "1") {
      pickedFile = await FilePicker.getFile(
          type: FileType.custom, allowedExtensions: ['jpg', 'png', 'mp4']);
      attachFile = File(pickedFile.path);
    } else if (_value == "2") {
      pickedFile = await _showCamera();
      attachFile = File(pickedFile);
    } else if (_value == "4") {
      attachFile = await FilePicker.getFile(
          type: FileType.custom, allowedExtensions: ['pdf']);
    } else {
      //to handle audio
      return;
    }

    final appDir = await syspaths.getApplicationDocumentsDirectory();
    fileName = path.basename(attachFile.path);
    savedImage = await attachFile.copy('${appDir.path}/$fileName');
    widget.onSelectImage(savedImage, attachFile.path);
    setState(() {
      _storedImage = attachFile;
      _filename = fileName;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: MediaQuery.of(context).size.width * 0.38,
      child: DropdownButton(
        isExpanded: true,
        underline: SizedBox(),
        icon: Icon(Icons.attachment, color: Colors.green),
        hint: Text("Attach Files", style: TextStyle(color: Colors.green)),
        items: [
          DropdownMenuItem(
            value: "1",
            child: Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: <Widget>[
                Icon(Icons.image),
//                SizedBox(
//                  width: MediaQuery.of(context).size.width*0.009,
//                ),
                Text(
                  "Image/Video",
                ),
              ],
            ),
          ),
          DropdownMenuItem(
            value: "2",
            child: Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: <Widget>[
                Icon(Icons.camera),
                Text(
                  "Camera",
                ),
              ],
            ),
          ),
          DropdownMenuItem(
            value: "4",
            child: Row(
              mainAxisAlignment: MainAxisAlignment.start,
              children: <Widget>[
                Icon(Icons.insert_drive_file),
                Text(
                  "File",
                ),
              ],
            ),
          ),
        ],
        onChanged: (value) {
          setState(() {
            var _value = value;
            _takePicture(_value);
          });
        },
      ),
    );
  }

  _showCamera() async {
    final cameras = await availableCameras();
    final camera = cameras.first;

    final result = await Navigator.push(
        context,
        MaterialPageRoute(
            builder: (context) => TakePicturePage(camera: camera)));

    return result;
  }
}


example/lib/TakePicturePage.dart
import 'package:camera/camera.dart';
import 'package:flutter/material.dart';
import 'package:path/path.dart';
import 'package:path_provider/path_provider.dart';

class TakePicturePage extends StatefulWidget {
  final CameraDescription camera;
  TakePicturePage({@required this.camera});

  @override
  _TakePicturePageState createState() => _TakePicturePageState();
}

class _TakePicturePageState extends State<TakePicturePage> {
  final fileName = DateTime.now().millisecondsSinceEpoch.toString();
  String vidPath;
  CameraController _cameraController;
  Future<void> _initializeCameraControllerFuture;
  int _selectedIndex = 0;
  bool _start = false;
  bool _isRec = false;
  void _onItemTapped(int index) {
    setState(() {
      _selectedIndex = index;
      if (_selectedIndex == 1) {
        _start = !_start;
      }
    });
  }

  @override
  void initState() {
    super.initState();

    _cameraController =
        CameraController(widget.camera, ResolutionPreset.medium);

    _initializeCameraControllerFuture = _cameraController.initialize();
    _fileInit();
  }

  void _fileInit() async {
    vidPath = join((await getTemporaryDirectory()).path, '${fileName}.mp4');
  }

  void _takePicture(BuildContext context) async {
    try {
      await _initializeCameraControllerFuture;

      if (_selectedIndex == 0) {
        final imgPath =
            join((await getTemporaryDirectory()).path, '${fileName}.png');
        await _cameraController.takePicture(imgPath);
        Navigator.pop(context, imgPath);
      } else {
        if (_start) {
          await _cameraController.startVideoRecording(vidPath);
          setState(() {
            _start = !_start;
            _isRec = !_isRec;
          });
        } else {
          _cameraController.stopVideoRecording();
          setState(() {
            _isRec = !_isRec;
          });
          Navigator.pop(context, vidPath);
        }
      }
    } catch (e) {
      print(e);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Stack(
        children: <Widget>[
          FutureBuilder(
            future: _initializeCameraControllerFuture,
            builder: (context, snapshot) {
              if (snapshot.connectionState == ConnectionState.done) {
                return CameraPreview(_cameraController);
              } else {
                return Center(child: CircularProgressIndicator());
              }
            },
          ),
          SafeArea(
            child: Align(
              alignment: Alignment.bottomCenter,
              child: Padding(
                padding: const EdgeInsets.all(8.0),
                child: FloatingActionButton(
                  backgroundColor: Colors.orange,
                  child: _selectedIndex == 1
                      ? _isRec == true
                          ? Icon(Icons.pause, color: Colors.white)
                          : Icon(Icons.play_arrow, color: Colors.white)
                      : Icon(Icons.camera, color: Colors.white),
                  onPressed: () {
                    _takePicture(context);
                  },
                ),
              ),
            ),
          ),
          _isRec == true
              ? SafeArea(
                  child: Container(
                    height: 30,
                    // alignment: Alignment.topLeft,
                    decoration: BoxDecoration(
                      color: Color(0xFFEE4400),
                      borderRadius: BorderRadius.circular(10),
                    ),
                    child: Padding(
                      padding: const EdgeInsets.all(8.0),
                      child: Text(
                        "REC",
                        style: TextStyle(
                            fontWeight: FontWeight.bold,
                            fontSize: 15,
                            color: Color(0xFFFAFAFA)),
                      ),
                    ),
                  ),
                )
              : SizedBox(
                  height: 0,
                )
        ],
      ),
      bottomNavigationBar: BottomNavigationBar(
        items: const <BottomNavigationBarItem>[
          BottomNavigationBarItem(
            icon: Icon(Icons.camera_alt),
            title: Text('Picture'),
          ),
          BottomNavigationBarItem(
            icon: Icon(Icons.videocam),
            title: Text('Video'),
          ),
        ],
        currentIndex: _selectedIndex,
        selectedItemColor: Colors.amber[800],
        onTap: _onItemTapped,
      ),
    );
  }

  @override
  void dispose() {
    _cameraController.dispose();
    super.dispose();
  }
}


example/lib/file_List_Preview.dart
import 'package:flutter/material.dart';
import 'package:open_file/open_file.dart';
import 'dart:io';
import 'dart:ui' as ui;

class FileListPreviewer extends StatefulWidget {
  // The list of Files
  final List<File> attachmentList;
  // The list of filepaths
  final List<String> filePaths;
  // The function to handle remove image functionality
  final Function removeImage;
  //The optional image for video files
  final String videoFileImage;
  //The width of the preview
  final double width;
  final key;
  //The width of the preview
  final double height;

  const FileListPreviewer({
    this.attachmentList,
    this.filePaths,
    this.videoFileImage = "assets/images/video.png",
    this.width = 300,
    this.height = 100,
    this.key,
    this.removeImage,
  });

  @override
  _FileListPreviewerState createState() => _FileListPreviewerState();
}

class _FileListPreviewerState extends State<FileListPreviewer> {
  List<File> _filePathList = [];

  bool _isLoading = false;
  //Method to remove the file attached
  void _removeImage(File pickedFile) {
    setState(() {
      _filePathList.remove(pickedFile); //removing image from the local list
    });
    widget.removeImage(
        pickedFile); // removing image from the list passed as data to file_list_previewer.
  }

  @override
  void initState() {
    super.initState();

    //If the input is filepaths,this condition converts filepaths to files else in other cases the input files are taken as input.
    // _filePathList is the local list of the data came as input,based on the input conditions checks and assigns data to the local list.

    if (widget.attachmentList == null && widget.filePaths != null) {
      widget.filePaths.forEach(
        (element) {
          File filepath = File(element);
          _filePathList.add(filepath);
        },
      );
    } else {
      _filePathList = widget.attachmentList;
    }
  }

  //Widget returns a listview builder

  @override
  Widget build(BuildContext context) {
    return MediaQuery(
      data: new MediaQueryData.fromWindow(ui.window),
      child: Directionality(
        textDirection: TextDirection.rtl,
        child: SingleChildScrollView(
          key: ValueKey('singlechildkey'),
          child: Container(
            key: ValueKey('containerkey'),
            height: widget.height,
            width: widget.width,
            child: _isLoading == true
                ? Center(
                    child: CircularProgressIndicator(
                      backgroundColor: Colors.blue,
                    ),
                  )
                :

                //ListView builder builds all the attached files which passed as input.
                ListView.builder(
                    key: ValueKey('listviewkey'),
                    scrollDirection: Axis.horizontal,
                    shrinkWrap: true,
                    itemCount: _filePathList != null ? _filePathList.length : 0,
                    itemBuilder: (BuildContext context, int index) {
                      // Finding the file extension to show appropriate thumbnail for that file type.
                      //File extension is found with substring of the filename
                      var extn = _filePathList[index].toString().substring(
                          _filePathList[index].toString().lastIndexOf('.'),
                          _filePathList[index].toString().length - 1);
                      return GestureDetector(
                        child: Stack(
                          children: [
                            //Conditions checks for image files and shows the image with help of image widget
                            (extn == ".jpg" || extn == ".jpeg" || extn == ".png"
                                ? new Container(
                                    width: 100.0,
                                    height: 100.0,
                                    decoration: new BoxDecoration(
                                      shape: BoxShape.rectangle,
                                      image: new DecorationImage(
                                          fit: BoxFit.fill,
                                          image: FileImage(_filePathList[index])
                                          //   image: new FileImage(attachFile),
                                          ),
                                    ),
                                  )
                                :
                                //Condition checks for video files and shows the thumbnail for video i.e, widget.videoFileImage
                                (extn == '.mp4'
                                    ? Container(
                                        width: 100.0,
                                        height: 100.0,
                                        //  child:Icon(Icons.videocam,size: 40,))
                                        child: Image(
                                          image:
                                              AssetImage(widget.videoFileImage),
                                          fit: BoxFit.scaleDown,
                                        ))
                                    :
                                    //In all other cases the default file icon card is shown.
                                    Material(
                                        type: MaterialType.transparency,
                                        child: SingleChildScrollView(
                                          child: Container(
                                            width: 100.0,
                                            height: 100.0,
                                            decoration: new BoxDecoration(
                                              color: Colors.white,
                                            ),
                                            child: Row(
                                                mainAxisAlignment:
                                                    MainAxisAlignment.center,
                                                children: [
                                                  Text(
                                                    "File",
                                                    style: TextStyle(
                                                        fontWeight:
                                                            FontWeight.bold,
                                                        fontSize: 20),
                                                  ),
                                                  Icon(Icons.insert_drive_file)
                                                ]),
                                          ),
                                        ),
                                      ))),
                            Positioned(
                              top: 0,
                              right: 0,
                              child: widget.removeImage != null
                                  ? GestureDetector(
                                      onTap: () => _removeImage(_filePathList[
                                          index]), //calls the _removeImage function to remove the file from list.
                                      child: Card(
                                        elevation: 10,
                                        color: Colors.white,
                                        shape: CircleBorder(),
                                        child: Icon(
                                          Icons.clear,
                                          size: 18,
                                        ),
                                      ),
                                    )
                                  : SizedBox(),
                            ),
                          ],
                        ),
                        onTap: () async {
                          //OnTap the file is opened with appropriate filetype with  OpenFile  plugin.
                          OpenFile.open(_filePathList[index].path);
                        },
                      );
                    },
                  ),
          ),
        ),
      ),
    );
  }
}


lib/file_list_previewer.dart
library file_list_previewer;

import 'package:flutter/material.dart';
import 'package:open_file/open_file.dart';
import 'dart:io';
import 'dart:ui' as ui;

class FileListPreviewer extends StatefulWidget {
  // The list of Files
  final List<File> attachmentList;
  // The list of filepaths
  final List<String> filePaths;
  // The function to handle remove image functionality
  final Function removeImage;
  //The optional image for video files
  final String videoFileImage;
  //The width of the preview
  final double width;
  final key;
  //The width of the preview
  final double height;

  const FileListPreviewer({
    this.attachmentList,
    this.filePaths,
    this.videoFileImage = "assets/images/video.png",
    this.width = 300,
    this.height = 100,
    this.key,
    this.removeImage,
  });

  @override
  _FileListPreviewerState createState() => _FileListPreviewerState();
}

class _FileListPreviewerState extends State<FileListPreviewer> {
  List<File> _filePathList = [];

  bool _isLoading = false;
  //Method to remove the file attached
  void _removeImage(File pickedFile) {
    setState(() {
      _filePathList.remove(pickedFile); //removing image from the local list
    });
    widget.removeImage(
        pickedFile); // removing image from the list passed as data to file_list_previewer.
  }

  @override
  void initState() {
    super.initState();

    //If the input is filepaths,this condition converts filepaths to files else in other cases the input files are taken as input.
    // _filePathList is the local list of the data came as input,based on the input conditions checks and assigns data to the local list.

    if (widget.attachmentList == null && widget.filePaths != null) {
      widget.filePaths.forEach(
        (element) {
          File filepath = File(element);
          _filePathList.add(filepath);
        },
      );
    } else {
      _filePathList = widget.attachmentList;
    }
  }

  //Widget returns a listview builder

  @override
  Widget build(BuildContext context) {
    return MediaQuery(
      data: new MediaQueryData.fromWindow(ui.window),
      child: Directionality(
        textDirection: TextDirection.rtl,
        child: SingleChildScrollView(
          key: ValueKey('singlechildkey'),
          child: Container(
            key: ValueKey('containerkey'),
            height: widget.height,
            width: widget.width,
            child: _isLoading == true
                ? Center(
                    child: CircularProgressIndicator(
                      backgroundColor: Colors.blue,
                    ),
                  )
                :

                //ListView builder builds all the attached files which passed as input.
                ListView.builder(
                    key: ValueKey('listviewkey'),
                    scrollDirection: Axis.horizontal,
                    shrinkWrap: true,
                    itemCount: _filePathList != null ? _filePathList.length : 0,
                    itemBuilder: (BuildContext context, int index) {
                      // Finding the file extension to show appropriate thumbnail for that file type.
                      //File extension is found with substring of the filename
                      var extn = _filePathList[index].toString().substring(
                          _filePathList[index].toString().lastIndexOf('.'),
                          _filePathList[index].toString().length - 1);
                      return GestureDetector(
                        child: Stack(
                          children: [
                            //Conditions checks for image files and shows the image with help of image widget
                            (extn == ".jpg" || extn == ".jpeg" || extn == ".png"
                                ? new Container(
                                    width: 100.0,
                                    height: 100.0,
                                    decoration: new BoxDecoration(
                                      shape: BoxShape.rectangle,
                                      image: new DecorationImage(
                                          fit: BoxFit.fill,
                                          image: FileImage(_filePathList[index])
                                          //   image: new FileImage(attachFile),
                                          ),
                                    ),
                                  )
                                :
                                //Condition checks for video files and shows the thumbnail for video i.e, widget.videoFileImage
                                (extn == '.mp4'
                                    ? Container(
                                        width: 100.0,
                                        height: 100.0,
                                        //  child:Icon(Icons.videocam,size: 40,))
                                        child: Image(
                                          image:
                                              AssetImage(widget.videoFileImage),
                                          fit: BoxFit.scaleDown,
                                        ))
                                    :
                                    //In all other cases the default file icon card is shown.
                                    Material(
                                        type: MaterialType.transparency,
                                        child: SingleChildScrollView(
                                          child: Container(
                                            width: 100.0,
                                            height: 100.0,
                                            decoration: new BoxDecoration(
                                              color: Colors.white,
                                            ),
                                            child: Row(
                                                mainAxisAlignment:
                                                    MainAxisAlignment.center,
                                                children: [
                                                  Text(
                                                    "File",
                                                    style: TextStyle(
                                                        fontWeight:
                                                            FontWeight.bold,
                                                        fontSize: 20),
                                                  ),
                                                  Icon(Icons.insert_drive_file)
                                                ]),
                                          ),
                                        ),
                                      ))),
                            Positioned(
                              top: 0,
                              right: 0,
                              child: widget.removeImage != null
                                  ? GestureDetector(
                                      onTap: () => _removeImage(_filePathList[
                                          index]), //calls the _removeImage function to remove the file from list.
                                      child: Card(
                                        elevation: 10,
                                        color: Colors.white,
                                        shape: CircleBorder(),
                                        child: Icon(
                                          Icons.clear,
                                          size: 18,
                                        ),
                                      ),
                                    )
                                  : SizedBox(),
                            ),
                          ],
                        ),
                        onTap: () async {
                          //OnTap the file is opened with appropriate filetype with  OpenFile  plugin.
                          OpenFile.open(_filePathList[index].path);
                        },
                      );
                    },
                  ),
          ),
        ),
      ),
    );
  }
}
